### PR TITLE
Moving domain exclusion logic to options

### PR DESCRIPTION
**Pull request by sandeshp**
This is a tenetative PR to be treated as Work In Progress. Points to note/discuss before finalizing -

* ListenerHandler was made abstract, forcing subclasses to define OnStartActivity.
The default impementation of just setting a sampler didnt help as the sampler needed the URI that has to be extracted in custom way.
* For this to be truly generic, the input param of the func may be not just Uri, but a bigger struct with more possibilities. However, I dont want to make this a generic object, as the individual listeners should have a hard protocol on what to pass in to get a sampler.
* The HttpInListener for Microsoft.AspNetCore did not have the exclusion for zipkin.azure... domain, hence initialized with a function hardcoded to return AlwaysSample in Startup.cs. Confirm/ discuss.
* Cleanup - Some code lines i commented out for now need to the cleaned before final merge (self-reminder) 
* I had to install Asnnet.Http Extensions to use GetDisplayUrl , which was just to pass into the sampler func, but don't think I needed it as the func anyway returns AlwaysSample. Discuss.

### CODE REVIEW

File: src/OpenCensus.Collector.Dependencies/DependenciesCollectorOptions.cs [hunk 1]
```diff
@@ -16,10 +16,30 @@
...
     /// </summary>
     public class DependenciesCollectorOptions
     {
+        private static Func<Uri, ISampler> defaultSampler = (uri) => { return uri.ToString().Contains("zipkin.azurewebsites.net") ? Samplers.NeverSample : Samplers.AlwaysSample; };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DependenciesCollectorOptions"/> class.
+        /// </summary>
+        /// <param name="sampler">Custom sampling function, if any</param>
+        public DependenciesCollectorOptions(Func<Uri, ISampler> sampler = null)
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Dependencies/DependenciesCollectorOptions.cs]: perhaps this func can take an `HttpRequest` object as an argument so one define a sampler based on cookies, not the url.

File: src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj [hunk 2]
```diff
@@ -33,6 +33,7 @@
 
   <ItemGroup>
     <ProjectReference Include="..\OpenCensus.Abstractions\OpenCensus.Abstractions.csproj" />
+    <PackageReference Include="Microsoft.AspNetCore.Http.Extensions" Version="2.0.0" />
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj]: I'm curious why this is needed?

File: src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs [hunk 3]
```diff
@@ -62,7 +64,7 @@ public override void OnStartActivity(Activity activity, object payload)
             string path = (request.PathBase.HasValue || request.Path.HasValue) ? (request.PathBase + request.Path).ToString() : "/";
 
             ISpan span = null;
-            this.Tracer.SpanBuilderWithRemoteParent(path, SpanKind.Server, ctx).SetSampler(this.Sampler).StartScopedSpan(out span);
+            this.Tracer.SpanBuilderWithRemoteParent(path, SpanKind.Server, ctx).SetSampler(this.Sampler(new Uri(request.GetDisplayUrl()))).StartScopedSpan(out span);
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.AspNetCore/Implementation/HttpInListener.cs]: `sampler` may return `null` which I think should indicate to use default. So I'd suggest to have this logic implemented here - check result for `null` and fall back to the default one

File: src/OpenCensus.Collector.Dependencies/DependenciesCollectorOptions.cs [hunk 4]
```diff
@@ -16,10 +16,30 @@
...
+    using System;
+    using OpenCensus.Trace;
+    using OpenCensus.Trace.Sampler;
+
     /// <summary>
     /// Options for dependencies collector.
     /// </summary>
     public class DependenciesCollectorOptions
     {
+        private static Func<Uri, ISampler> defaultSampler = (uri) => { return uri.ToString().Contains("zipkin.azurewebsites.net") ? Samplers.NeverSample : Samplers.AlwaysSample; };
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Dependencies/DependenciesCollectorOptions.cs]: I'd suggest to return `Samplers.NeverSample` in case of `zipkin` url and `null` otherwise. Basically only return something meaningful when you know how to handle this specific span.

File: src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs [hunk 5]
```diff
@@ -16,16 +16,17 @@
 
 namespace OpenCensus.Collector.Implementation.Common
 {
+    using System;
     using System.Diagnostics;
     using OpenCensus.Trace;
 
-    internal class ListenerHandler
+    internal abstract class ListenerHandler
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs:24]: I feel it may be better to simply have two copies of implementation instead of abstracting it away. So basically copu/paste code.

**Review by SergeyKanzhelev — CHANGES_REQUESTED** [review]
Great start!

File: src/OpenCensus.Collector.Dependencies/OpenCensus.Collector.Dependencies.csproj [hunk 6]
```diff
@@ -32,4 +32,10 @@
   <ItemGroup Condition="'$(TargetFramework)' == 'net46'">
     <Reference Include="System.Net.Http" />
   </ItemGroup>
+
+  <ItemGroup>
+    <Reference Include="Microsoft.AspNetCore.Http.Abstractions">
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Dependencies/OpenCensus.Collector.Dependencies.csproj]: you don't need this. Please remove

File: src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs [hunk 7]
```diff
@@ -16,16 +16,18 @@
...
+    using System.Net.Http;
     using OpenCensus.Trace;
 
-    internal class ListenerHandler
+    internal abstract class ListenerHandler
     {
         protected readonly ITracer Tracer;
 
-        protected readonly ISampler Sampler;
+        protected readonly Func<HttpRequestMessage, ISampler> Sampler;
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs]: Perhaps call the field `SamplerFactory` to highlight the semantics of it.

File: test/TestApp.AspNetCore.2.0/Startup.cs [hunk 8]
```diff
@@ -49,7 +49,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddSingleton<ISampler>(Samplers.AlwaysSample);
             services.AddSingleton<RequestsCollectorOptions>(new RequestsCollectorOptions());
             services.AddSingleton<RequestsCollector>();
-            services.AddSingleton<DependenciesCollectorOptions>(new DependenciesCollectorOptions());
+            services.AddSingleton<DependenciesCollectorOptions>(new DependenciesCollectorOptions((uri) => { return Samplers.AlwaysSample; }));
```
Comment (SergeyKanzhelev) [inline test/TestApp.AspNetCore.2.0/Startup.cs]: you don't need to pass it here. It is already set up as a default sampler. Having it here may be confusing

File: src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs [hunk 9]
```diff
@@ -34,10 +36,11 @@ public ListenerHandler(string sourceName, ITracer tracer, ISampler sampler)
 
         public string SourceName { get; }
 
-        public virtual void OnStartActivity(Activity activity, object payload)
-        {
-            this.Tracer.SpanBuilder(activity.OperationName).SetSampler(this.Sampler).StartScopedSpan();
-        }
+        public abstract void OnStartActivity(Activity activity, object payload);
+
+        // {
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs]: just delete those lines, don't keep in comments

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs [hunk 10]
```diff
@@ -40,10 +41,14 @@ public class DependenciesCollector : IDisposable
...
         {
             this.diagnosticSourceSubscriber = new DiagnosticSourceSubscriber(
-                new Dictionary<string, Func<ITracer, ISampler, ListenerHandler>>()
+                new Dictionary<string, Func<ITracer, Func<HttpRequestMessage, ISampler>, ListenerHandler>>()
                 { { "HttpHandlerDiagnosticListener", (t, s) => new HttpHandlerDiagnosticListener(t, s, propagationComponent) } },
                 tracer,
-                sampler);
+                x =>
+                {
+                    ISampler s = options.CustomSampler(x);
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs]: catch exception here. If method failed with exception - default sampler should be used

File: src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj [hunk 11]
```diff
@@ -17,10 +17,6 @@
     <PackageRequireLicenseAcceptance>true</PackageRequireLicenseAcceptance>
   </PropertyGroup>
 
-  <ItemGroup>
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj:20]: should the same be done for dependencies project?

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,7 +7,12 @@ the release.
 
 ## Unreleased
 
-TBD
+- (Func returning) Custom sampler added to DependenciesCollector and RequestsCollector options
```
Comment (SergeyKanzhelev) [inline CHANGELOG.md]: I'd suggest to have something like:

> - OpenCensus.Collectors.AspNetCore: Allow to supply custom sampler based on request properties using custom code. For instance filter out telemetry from specific path.
> - OpenCensus.Collectors.Dependencies: Allow to supply custom sampler based on request properties using custom code. For instance, filter out calls to Zipkin REST endpoint from the exporter.

**Review by SergeyKanzhelev — CHANGES_REQUESTED** [review]
thanks! Just a few updates needed

File: CHANGELOG.md (hunk 12 above)
Comment (sandeshp) [reply CHANGELOG.md]: Updated, let me know if you prefer to take off last 2 lines

File: src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj (hunk 11 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj:20]: If you mean PackageRequireLicenseAcceptance flag - it is true in both projects.
If you mean creating a separate subfolder and moving away from common, I kept the common thinking it may be useful if we have another listener based on HttpRequestMessage. We could still reuse code across listeners, just could not accommodate AspNetCore due to the different return type. Let me know if we want to eliminate common/ and just dup the code per listener type.

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs (hunk 10 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs]: Done. Q - If I put in a Debug.WriteLine(..) here in exception case, where would it emerge, would the caller be able to see it in event viewer ?

File: src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs (hunk 9 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs]: Done

File: test/TestApp.AspNetCore.2.0/Startup.cs (hunk 8 above)
Comment (sandeshp) [reply test/TestApp.AspNetCore.2.0/Startup.cs]: Done.

File: src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs (hunk 7 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Implementation.Common/ListenerHandler.cs]: Done

File: src/OpenCensus.Collector.Dependencies/OpenCensus.Collector.Dependencies.csproj (hunk 6 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Dependencies/OpenCensus.Collector.Dependencies.csproj]: Eliminated

File: CHANGELOG.md (hunk 12 above)
Comment (SergeyKanzhelev) [reply CHANGELOG.md]: Yes, please remove last two lines. We typically list public facing things to changelog, not implementation details

File: src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj (hunk 11 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj:20]: No I meant move files from `Common` folder to specific folder under the project folder

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs (hunk 10 above)
Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs]: It may be outside of a scope, but we will need to log error in Event Source https://github.com/census-instrumentation/opencensus-csharp/blob/develop/docs/error-logging.md in future,

File: src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs [hunk 13]
```diff
@@ -40,12 +41,16 @@ public class RequestsCollector : IDisposable
...
-                new Dictionary<string, Func<ITracer, ISampler, ListenerHandler>>()
+                new Dictionary<string, Func<ITracer, Func<HttpRequest, ISampler>, ListenerHandler>>()
                 {
                     { "Microsoft.AspNetCore", (t, s) => new HttpInListener(t, s, propagationComponent) },
                 },
                 tracer,
-                sampler);
+                x =>
+                {
+                    ISampler s = options.CustomSampler(x);
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.AspNetCore/RequestsCollector.cs]: I believe I commented here to handle exceptions thrown by custom sampler

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs [hunk 14]
```diff
@@ -40,10 +41,23 @@ public class DependenciesCollector : IDisposable
...
+                x =>
+                {
+                    ISampler s = null;
+                    try
+                    {
+                        s = options.CustomSampler(x);
+                    }
+                    catch (Exception)
+                    {
+                        s = null;
```
Comment (SergeyKanzhelev) [inline src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs:56]: please add `TODO:` to log error. (Error logging can be implemented as separate PR)

**Review by SergeyKanzhelev — CHANGES_REQUESTED** [review]
Few more changes needed before merging

File: src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj (hunk 11 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.AspNetCore/OpenCensus.Collector.AspNetCore.csproj:20]: Moved.

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs (hunk 10 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs]: Marked as TODO

File: CHANGELOG.md (hunk 12 above)
Comment (sandeshp) [reply CHANGELOG.md]: Removed

File: src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs (hunk 14 above)
Comment (sandeshp) [reply src/OpenCensus.Collector.Dependencies/DependenciesCollector.cs:56]: Added TODO

**Review by SergeyKanzhelev — APPROVED** [review]
